File: Cluster.py
import copy
import pandas as pd
import time

def generateClusters(buildings):		# jobs df not needed
	# filter through data to get clusters to work on individually rather than as a whole
	# just do it on the whole thing
	sets = []

	for b in buildings.index:
		curr = {b}
		for p in buildings.loc[b]['Worker1':'Worker6']:
			if not pd.isna(p):
				curr.add(p)
		existsIn = []
		for i in range(len(sets)):
			for s in curr:
				if s in sets[i] and i not in existsIn:
					existsIn.append(i)					# duplicate indices are NOT fine
		
		# print(len(sets), existsIn)
		while len(existsIn) > 0:
			curr = curr.union(sets[existsIn[-1]])				# must be added/removed from the back to keep sets[] indices intact
			sets.remove(sets[existsIn[-1]])
			existsIn.pop()
		sets.append(curr)

	sets = sorted(sets, key=lambda x: len(x))
	# print(sets)
	return sets

# state = [int, {'Building': Building()}]
def printState(curr):
	print('\nState:')
	max = 0
	if type(curr[0]) is int:
		max = 1
		print(curr[0])

	for key in curr[0+max]:
		print(key, curr[0+max][key],end=', ')
	# print(curr[1+max])

# can do individual clusters of permutations
class Cluster:
	def __init__(self, buildings,jobs,qbmap,qpmap):
		self.buildings = buildings
		self.jobs = jobs
		self.qbmap = qbmap
		self.qpmap = qpmap
		
		self.max = [0, {}]
		self.done = False

		# print(self.qbmap, self.qpmap)
	def __str__(self):
		if not self.done:
			self.getMax()
		# print value
		ret = 'Val ' + str(self.max[0]) + '\nBuildings: {'
		
		# print buildings
		buildings = copy.deepcopy(self.max[1])
		reversed(buildings)
		for key in buildings:
			ret += key + ' : ' + str(buildings[key]) + ', '
		ret = ret[:-2] + '}\n'
		
		# if len(self.max[2]) > 0:
		# 	ret += 'People: {'

		# 	# print people
		# 	people = copy.deepcopy(self.max[2])
		# 	for key in people:
		# 		ret += key + ' : ' + people[key] + ', '
		# 	ret = ret[:-2] + '}\n'
		
		return f"{ret}"

	class Building:
		# input: number of buildings
		def __init__(self, buildings, num):
			self.buildings = buildings
			self.jobs = [set() for i in range(int(num))]
		def __str__(self):
			return f"{self.jobs}"
		def __len__(self):
			return len(self.jobs)
		

		# input: job that we want to assign
		# returns True if success, False if fail
		# no need to error check since jobs will be selected by those whose workplace is here
		def insertJob(self, j: str) -> bool:
			for i in range(len(self.jobs)):
				if j not in self.jobs[i]:
					self.jobs[i].add(j)
					return True
			return False
		
		# input: job that we want to assign
		# returns True if success, False if fail
		def removeJob(self, j: str) -> bool:
			for i in range(len(self.jobs)-1, -1, -1):
				if j in self.jobs[i]:
					self.jobs[i].remove(j)
					return True
			return False
		
		def getCPS(self, name) -> int:
			sum = 0
			for i in self.jobs:
				# print(name, i)
				# print(buildings.loc[name])

				if len(i) < int(self.buildings.loc[name]['Capacity']):
					sum += len(i) * int(self.buildings.loc[name]['Multiplier'][1])
				elif len(i) > int(self.buildings.loc[name]['Capacity']):
					print("Error: more jobs than capacity allowed at:", name)
					exit(1)
				else:
					sum += int(self.buildings.loc[name]['MaxCPS'])
			return sum

	# do the jobs that only have one workplace first, or if we have a one-to-one quantity
	# will automatically update self.max
	def filterEasy(self, builds, people):
		# preload the buildings with the no-brainer options
		# append the no-brainers to the phash at the end
		ez_people = {}		# dict for all no-brainers
		poplist = []
		for i in people:
			workplaces = self.jobs.loc[i[:-1]]['Workplace1':'Workplace3']
			if int(self.jobs.loc[i[:-1]]['Options']) == 1:
				ez_people[i] = workplaces[0]
				# self.max[2][i] = ez_people[i]
				
				builds[ez_people[i]].insertJob(i[:-1])
				self.max[1][ez_people[i]].insertJob(i[:-1])
				poplist.append(i)
				
		for i in poplist:
			people.pop(i)
		return ez_people
	
	# if qperson == qbuildings they can be assigned, do those
	def filterFit(self, builds, qpmap):
		# count = []
		poplist = []
		for i in qpmap:
			count = []
			workplaces = self.jobs.loc[i]['Workplace1':'Workplace3'].dropna().to_list()
			for w in workplaces:					# get the number of buildings (quantity) that this person can go to
				count.append(len(builds[w])) if w in builds else count.append(0)
			print(i, count)
			if sum(count) == qpmap[i]:
				for c in range(len(count)):
					if workplaces[c] in builds:
						while count[c] > 0:
							builds[workplaces[c]].insertJob(i)
							self.max[1][workplaces[c]].insertJob(i)
							count[c] -= 1
							qpmap[i] -= 1
				poplist.append(i)

		for i in poplist:
			qpmap.pop(i)
		# return ez_people

	def calcCPS(self, bdict) -> int:
		sum = 0
		for i in bdict:
			sum += bdict[i].getCPS(i)
		return sum

	# iterate through keys and get a value to save
	def recursion(self, blist, plist, curr, iter=0, count=[0,0]):
		# print('Function begin:')
		# base case: we have reached the end of the people
		if iter == len(plist):
			count[0] += 1
			if count[0] % 1000 == 0 or count[0] == count[1]:
				print('Generated', count[0], 'of', count[1], 'permutations')
			
			val = self.calcCPS(curr[0])
			if val > self.max[0]:
				self.max[0] = val
				self.max[1] = copy.deepcopy(curr[0])
				# self.max[2] = dict(curr[1])
			return
		
		# iterative step: set one of the jobs to a building, then recursively iterate through the rest
		
		workplaces = self.jobs.loc[plist[iter][:-1]]['Workplace1':'Workplace3'].dropna().to_list()
		# workplaces = [''] + workplaces				# assume every person will not have a place to go

		for w in workplaces:
			if not pd.isna(w):
			
				if w in blist and curr[0][w].insertJob(plist[iter][:-1]):			# add job to building. make sure assignment worked
					curr[1][plist[iter]] = w

				# printState(curr)
				
				# go to next level
				self.recursion(blist, plist, curr, iter+1, count)

				if w in blist:
					curr[0][w].removeJob(plist[iter][:-1])
					curr[1][plist[iter]] = ''


	# input: building_data, job_data, building_quantity, people_quantity
	def getMax(self):
		if self.done:
			return self.max
		
		# set up data
		builds = {}
		for b in self.qbmap.keys():
			temp = self.Building(self.buildings, self.qbmap[b])
			builds[b] = temp
		self.max[1] = copy.deepcopy(builds)

		qpmap = dict(self.qpmap)
		self.filterFit(builds, qpmap)

		people = {}
		for p in qpmap.keys():
			while qpmap[p] > 0:
				people[p+str(int(qpmap[p]))] = ''
				qpmap[p] -= 1

		ez_people = self.filterEasy(builds, people)		# filter the easy ones out

		# I'm just going to use dicts for people and buildings
		curr = [dict(builds),dict(people)]
		
		total = 1
		for i in people:
			num = float(self.jobs.loc[i[:-1]]['Options'])
			# num += 1.0				# +1 for empty
			total *= num
		print('Need to generate',total,'cases.')
		count = [0, total]				# show a load time
		# put it into recursion
		self.start_time = time.time()
		self.recursion(tuple(builds.keys()), tuple(people.keys()), curr, count=count)
		print('Finished recursion')
		print('Total time:', time.time()-self.start_time)

		# for i in ez_people:
		# 	self.max[2][i] = ez_people[i]

		# printState(self.max)
		self.done = True
		# input('.')

		return self.max

File: Cluster.cpp
#include "Cluster.h"
#include <bits/stdc++.h>
#include "DF_Headers.h"
#include "Prints.h"
using namespace std;

// default constructor
// needed for default creation in builds map, but there are branches in functions (that return 0) just in case
Cluster::Building::Building() {
	info = nullptr;
	// workers.clear()
	// fullCPS = false;
}

// input is address of the dataframe row and quantity
Cluster::Building::Building(vector<string>* building, int num = 1) : info(building) {
	workers.resize(num);	// will create num number of empty sets
}

// print workers
string Cluster::Building::str() const {
	string ans = "[";
	for (unsigned int i = 0; i < workers.size(); i++) {
		if (i > 0) {
			ans.append(", ");
		}
		ans.append(setToStr(workers.at(0)));
	} ans.append("]");
	return ans;
}
// return number of this type of building (workers.size())
unsigned int Cluster::Building::len() const {
	return workers.size();
}
// clear all sets in workers
void Cluster::Building::clear() {
	for (unsigned i = 0; i < workers.size(); i++) {
		workers.at(i).clear();
	}
}


// true if job can and was added, false if not
bool Cluster::Building::insertJob(const string& j) {
	if (info == nullptr)
		return false;

	unsigned int prev_size;
	for (unsigned i = 0; i < workers.size(); i++) {
		prev_size = workers.at(i).size();
		workers.at(i).insert(j);
		if (prev_size != workers.at(i).size())
			return true;
	}
	return false;
}
// true if job exists and was removed, false if not
bool Cluster::Building::removeJob(const string& j) {
	if (info == nullptr)
		return false;

	unsigned int prev_size;
	for (unsigned i = 0; i < workers.size(); i++) {
		prev_size = workers.at(i).size();
		workers.at(i).erase(j);
		if (prev_size != workers.at(i).size())
			return true;
	}
	return false;
}

// int Cluster::Building::totalMult() const {
// 	if (!fullCPS) {
// 		return 1;
// 	}
// }
// calculate all building CPS based on number of workers
int Cluster::Building::getCPS() const {
	if (info == nullptr)
		return 0;

	int sum = 0, multiplier;
	for (auto& i: workers) {
		// if we have a full number of people, use maxcps
		if (i.size() == unsigned(stoi(info->at(Capacity)))) {
			sum += stoi(info->at(MaxCPS));
		}
		else {
			multiplier = stoi(info->at(Multiplier).substr(1));
			sum += multiplier * i.size();
		}
	}
	return sum;
}

// -------------------

Cluster::Cluster(map<string, vector<string>> *buildings, map<string, vector<string>> *jobs, const map<string, int>& qbmap, const map<string, int>& qpmap) :
buildings(buildings), jobs(jobs), qbmap(qbmap), qpmap(qpmap) {
	maxVal = 0;
	count = 0;
	total = 0;
	exponent = 0;

	done = false;
}

// print max and config
string Cluster::str() {
	if (!done) {
		getMax();
	}

	string ans = "Max: ";
	ans.append(to_string(getMax()));
	ans.push_back('\n');
	ans.append(configStr());
	ans.push_back('\n');

	return ans;
}

// assign the jobs that only have one workplace
// make changes to parameters
void Cluster::filterEasy(map<string, Building>& builds, vector<string>& people) {
	vector<string> newp;
	string currPerson;
	vector<string> currRow;
	for (unsigned i = 0; i < people.size(); i++) {
		currPerson = people.at(i).substr(0, people.at(i).size()-1);
		currRow = jobs->at(currPerson);
		
		// needs to have only 1 workplace.
		// should have a spot guaranteed
		if (!(stoi(currRow.at(Options)) == 1 && builds[currRow.at(Workplace1)].insertJob(currPerson))) {
			newp.push_back(people.at(i));
		}
	}
	people = newp;
}
// assign jobs where we have a one-to-one quantity between the job and their buildings
void Cluster::filterFit(map<string, Building>& builds, map<string,int>& qpmap) {
	int sum;
	vector<int>curr(3);
	vector<string> currRow;
	for (auto i: qpmap) {
		sum = 0;
		curr.resize(3,0);
		currRow = jobs->at(i.first);
		// get number of buildings
		for (unsigned j = 0; j < 3 && !currRow.at(j).empty(); j++) {
			curr.at(j) += builds[currRow.at(j+Workplace1)].len();
			sum += curr.at(j);
		}

		// if the numbers match, put them in buildings and remove the job
		if (sum == i.second) {
			for (unsigned j = 0; j < 3 && !currRow.at(j).empty(); j++) {
				while (curr.at(j)) {
					builds[currRow.at(j+Workplace1)].insertJob(i.first);
					curr.at(j)--;
				}
			}
			qpmap.erase(i.first);
		}
	}
}
// iterate through all of the buildings and get their CPS's
int Cluster::calcCPS(const map<string, Building>& bdict) const {
	int sum = 0;
	for (auto& i: bdict) {
		sum += i.second.getCPS();
	}
	return sum;
}
void printState(map<string, Cluster::Building>& currBuilds) {
	cout << "State:\n{";
	for (map<string,Cluster::Building>::iterator i = currBuilds.begin(); i != currBuilds.end(); i++) {
		if (i != currBuilds.begin()) {
			cout << ", ";
		}
		cout << i->first << " : " << i->second.str();
	}
	cout << "}\n";
}

// iterate through keys and get a value to save
// also uses this->count to keep track of progress (how much the algorithm has done)
void Cluster::recursion(map<string, Building>& currBuilds, const vector<string>& plist, unsigned iter = 0) {
	// base case: we have reached the end of the people
	if (iter == plist.size()) {
		// progress record
		this->count++;
		if (this->count % 100000 == 0 || (this->count == this->total && this->ce >= this->exponent)) {
			if (count > 1000000000) {
				this->ce += 9;
				count = 0;
			}
			cout << "Generated " << this->count << "e" << this->ce << " of " << this->total << "e" << exponent << " permutations.\n";
		}
		
		// see if current config has highest value
		int val = calcCPS(currBuilds);
		if (val > this->maxVal) {
			this->maxVal = val;
			this->maxMap = currBuilds;		// should be an easy transfer? but not efficient in terms of space probably
		}
		return;
	}

	vector<string> workplaces = {};		// can add "" if there can be person with no job
	string currJob = plist[iter].substr(0, plist[iter].size()-1);	// remove the id number when adding to building
	vector<string> currRow = jobs->at(currJob);
	for (unsigned i = Workplace1; i <= Workplace3 && !currRow.at(i).empty(); i++) {
		if (currBuilds.find(currRow[i]) != currBuilds.end())
		workplaces.push_back(currRow.at(i));
	}

	for (auto w: workplaces) {
		// make sure we can actually have the building and can add it
		if (currBuilds[w].insertJob(currJob)) {
			recursion(currBuilds, plist, iter+1);

			currBuilds[w].removeJob(currJob);
		}
	}

}
int Cluster::getMax() {
	if (done) {
		return maxVal;
	}

	// set up data
	map<string, Building> builds;
	vector<string> *row;
	for (auto b: this->qbmap) {
		row = &buildings->at(b.first);
		Building temp(row, b.second);
		builds[b.first] = temp;
	}

	map<string, int> qpmap = this->qpmap;
	filterFit(builds, qpmap);		// useful to make obvious choices

	vector<string> people;
	for (map<string, int>::iterator i = qpmap.begin(); i != qpmap.end(); i++) {
		while (i->second > 0) {
			people.push_back(i->first + to_string(i->second));
			qpmap[i->first]--;
		}
	}

	filterEasy(builds, people);		// useful to make obvious choices
	cout << "Trimmed. " << people.size() << " people remain." << endl;

	// calculate how long it will take
	total = 1;
	long num;
	for (auto i: people) {
		if (jobs->find(i.substr(0, i.size()-1)) == jobs->end()) {
			cout << "Error: " << i.substr(0, i.size()-1) << " does not exist in jobs.csv." << endl;
			exit(1);
		}
		num = stoul(jobs->at(i.substr(0, i.size()-1)).at(Options));
		total *= num;
		if (total > 1000000000) {
			exponent += 9;
			total /= 1000000000;
		}
		cout << "Total: " << total << "e" << exponent << endl;
	}
	cout << "Need to generate " << total << "e" << exponent << " cases.\n";

	// auto start = time(0);
	recursion(builds, people);
	cout << "Finished recursion.\n";
	// cout << "Total time: " << time(0)-start << "\n";
	cout << endl;

	done = true;

	return maxVal;
}

string Cluster::configStr() {
	if (!done) {
		getMax();
	}

	string ans;
	// can't use printMap(), so gotta write it normally
	ans = "{";
	for (map<string, Building>::iterator i = maxMap.begin(); i != maxMap.end(); i++) {
		if (i != maxMap.begin())
			ans.append(", ");
		ans.append(i->first);
		ans.append(" : ");
		ans.append(i->second.str());
	}
	ans.append("}\n");

	return ans;
}


// create a vector of sets that have all connected buildings and people
vector<set<string>> generateClusters(const map<string, vector<string>>& buildings) {
	vector<set<string>> ans;

	// for each building
	// get the jobs that can go there and throw them all in a set
	// if we encounter a job that we have already seen, union the sets
	// finally, push to ans
	set<string> temp;
	for (auto& i: buildings) {
		temp.clear();

		temp.insert(i.second[Building_name]);
		for (unsigned j = Worker1; j <= Worker6; j++) {
			if (!i.second[j].empty()) {
				temp.insert(i.second[j]);
			}
		}
		// cout << setToStr(temp) << endl;

		for (unsigned k = 0; k < ans.size(); k++) {
			for (auto& j: temp) {
				if (ans.at(k).find(j) != ans.at(k).end()) {
					temp.merge(ans.at(k));
					ans.at(k).clear();
					continue;
				}
			}
		}
		ans.push_back(temp);
	}

	// clean up ans by removing all empty sets used
	vector<set<string>> ret;
	for (unsigned i = 0; i < ans.size(); i++) {
		if (!ans.at(i).empty())
			ret.push_back(ans.at(i));
	}

	// sort vector by set size
	auto cmp = [](const set<string>& s1, const set<string>& s2) { return s1.size() < s2.size(); };
	sort(ret.begin(), ret.end(), cmp);

	cout << ret.size() << " clusters generated.\n";
	
	return ret;
}
